### ISSUE TITLE
Graphite query does not support * within a  {...} block

### ISSUE THREAD

**Issue by JGShaw** [VictoriaMetrics/VictoriaMetrics#952]
When querying using graphite syntax, putting a * with a {...} causes no metrics to be returned. Following exactly the same steps with a basic Graphite install has the expected behaviour show here.

**To Reproduce**
Add the following metrics:
```
one1.foo 0
one2.bar 0
two3.baz 0
```

Query using expand (find has a the same bug as well, so it appears to apply to all graphite queries):
```
curl -g 'localhost:8428/metrics/expand/?query={one*}'
```

It will return:
```
[]
```

**Expected behavior**
it should expand the query to match `one1` and `one2` and give the result:
```
["one1.","one2."]
```

**Version**
v.1.49.0

**Used command-line flags**
flag{name="bigMergeConcurrency", value="0"} 1
flag{name="csvTrimTimestamp", value="1ms"} 1
flag{name="dedup.minScrapeInterval", value="0s"} 1
flag{name="deleteAuthKey", value="secret"} 1
flag{name="denyQueriesOutsideRetention", value="false"} 1
flag{name="dryRun", value="false"} 1
flag{name="enableTCP6", value="false"} 1
flag{name="envflag.enable", value="false"} 1
flag{name="envflag.prefix", value=""} 1
flag{name="finalMergeDelay", value="30s"} 1
flag{name="forceFlushAuthKey", value="secret"} 1
flag{name="forceMergeAuthKey", value="secret"} 1
flag{name="fs.disableMmap", value="false"} 1
flag{name="graphiteListenAddr", value=":2003"} 1
flag{name="graphiteTrimTimestamp", value="1s"} 1
flag{name="http.connTimeout", value="2m0s"} 1
flag{name="http.disableResponseCompression", value="false"} 1
flag{name="http.idleConnTimeout", value="1m0s"} 1
flag{name="http.maxGracefulShutdownDuration", value="7s"} 1
flag{name="http.pathPrefix", value=""} 1
flag{name="http.shutdownDelay", value="0s"} 1
flag{name="httpAuth.password", value="secret"} 1
flag{name="httpAuth.username", value=""} 1
flag{name="httpListenAddr", value=":8428"} 1
flag{name="import.maxLineLen", value="104857600"} 1
flag{name="influx.maxLineSize", value="262144"} 1
flag{name="influxListenAddr", value=""} 1
flag{name="influxMeasurementFieldSeparator", value="_"} 1
flag{name="influxSkipMeasurement", value="false"} 1
flag{name="influxSkipSingleField", value="false"} 1
flag{name="influxTrimTimestamp", value="1ms"} 1
flag{name="insert.maxQueueDuration", value="1m0s"} 1
flag{name="loggerDisableTimestamps", value="false"} 1
flag{name="loggerErrorsPerSecondLimit", value="0"} 1
flag{name="loggerFormat", value="default"} 1
flag{name="loggerLevel", value="INFO"} 1
flag{name="loggerOutput", value="stderr"} 1
flag{name="loggerWarnsPerSecondLimit", value="0"} 1
flag{name="maxConcurrentInserts", value="8"} 1
flag{name="maxInsertRequestSize", value="33554432"} 1
flag{name="maxLabelsPerTimeseries", value="30"} 1
flag{name="memory.allowedBytes", value="0"} 1
flag{name="memory.allowedPercent", value="60"} 1
flag{name="metricsAuthKey", value="secret"} 1
flag{name="opentsdbHTTPListenAddr", value=""} 1
flag{name="opentsdbListenAddr", value=""} 1
flag{name="opentsdbTrimTimestamp", value="1s"} 1
flag{name="opentsdbhttp.maxInsertRequestSize", value="33554432"} 1
flag{name="opentsdbhttpTrimTimestamp", value="1ms"} 1
flag{name="pprofAuthKey", value="secret"} 1
flag{name="precisionBits", value="64"} 1
flag{name="promscrape.config", value=""} 1
flag{name="promscrape.config.dryRun", value="false"} 1
flag{name="promscrape.config.strictParse", value="false"} 1
flag{name="promscrape.configCheckInterval", value="0s"} 1
flag{name="promscrape.consulSDCheckInterval", value="30s"} 1
flag{name="promscrape.disableCompression", value="false"} 1
flag{name="promscrape.disableKeepAlive", value="false"} 1
flag{name="promscrape.discovery.concurrency", value="100"} 1
flag{name="promscrape.discovery.concurrentWaitTime", value="1m0s"} 1
flag{name="promscrape.dnsSDCheckInterval", value="30s"} 1
flag{name="promscrape.dockerswarmSDCheckInterval", value="30s"} 1
flag{name="promscrape.dropOriginalLabels", value="false"} 1
flag{name="promscrape.ec2SDCheckInterval", value="1m0s"} 1
flag{name="promscrape.eurekaSDCheckInterval", value="30s"} 1
flag{name="promscrape.fileSDCheckInterval", value="30s"} 1
flag{name="promscrape.gceSDCheckInterval", value="1m0s"} 1
flag{name="promscrape.kubernetesSDCheckInterval", value="30s"} 1
flag{name="promscrape.maxDroppedTargets", value="1000"} 1
flag{name="promscrape.maxScrapeSize", value="16777216"} 1
flag{name="promscrape.openstackSDCheckInterval", value="30s"} 1
flag{name="promscrape.streamParse", value="false"} 1
flag{name="promscrape.suppressDuplicateScrapeTargetErrors", value="false"} 1
flag{name="promscrape.suppressScrapeErrors", value="false"} 1
flag{name="relabelConfig", value=""} 1
flag{name="retentionPeriod", value="1"} 1
flag{name="search.cacheTimestampOffset", value="5m0s"} 1
flag{name="search.disableCache", value="false"} 1
flag{name="search.latencyOffset", value="30s"} 1
flag{name="search.logSlowQueryDuration", value="5s"} 1
flag{name="search.maxConcurrentRequests", value="4"} 1
flag{name="search.maxExportDuration", value="720h0m0s"} 1
flag{name="search.maxLookback", value="0s"} 1
flag{name="search.maxPointsPerTimeseries", value="30000"} 1
flag{name="search.maxQueryDuration", value="30s"} 1
flag{name="search.maxQueryLen", value="16384"} 1
flag{name="search.maxQueueDuration", value="10s"} 1
flag{name="search.maxStalenessInterval", value="0s"} 1
flag{name="search.maxTagKeys", value="secret"} 1
flag{name="search.maxTagValueSuffixesPerSearch", value="100000"} 1
flag{name="search.maxTagValues", value="100000"} 1
flag{name="search.maxUniqueTimeseries", value="300000"} 1
flag{name="search.minStalenessInterval", value="0s"} 1
flag{name="search.resetCacheAuthKey", value="secret"} 1
flag{name="search.treatDotsAsIsInRegexps", value="false"} 1
flag{name="selfScrapeInstance", value="self"} 1
flag{name="selfScrapeInterval", value="0s"} 1
flag{name="selfScrapeJob", value="victoria-metrics"} 1
flag{name="smallMergeConcurrency", value="0"} 1
flag{name="snapshotAuthKey", value="secret"} 1
flag{name="storageDataPath", value="victoria-metrics-data"} 1
flag{name="tls", value="false"} 1
flag{name="tlsCertFile", value=""} 1
flag{name="tlsKeyFile", value="secret"} 1
flag{name="version", value="false"} 1

**Comment by valyala** [comment]
@JGShaw , the issue must be fixed in the commit ba8ac08739abad55a59320202e5271ea96e539ab . Could you build VictoriaMetrics from this commit according to [these instructions](https://victoriametrics.github.io/#how-to-build-from-sources) and verify whether this fixes the bug?

**Comment by JGShaw** [comment]
@valyala I ran the development build on that commit and it has fixed the issue! Thanks a lot for fixing so quickly!

Do you know how long roughly it will take until this commit makes it way into a release?

**Comment by valyala** [comment]
> Do you know how long roughly it will take until this commit makes it way into a release?

The next release (v1.50.0) should be published on the next week. I'll post a notification here when the release will be ready.

**Comment by valyala** [comment]
The bugfix has been included in [v1.50.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases) release. Closing the issue as fixed.

**Comment by marcuzh** [comment]
Hi @valyala

Apologies, I'm a bit of a VictoriaMetrics n00b so I'll try my best to phrase this in a helpful way :-)
I didn't want to raise it as a separate bug just yet as I don't think I have the right evidence needed.

Myself and @JGShaw are now running 1.50.1 in production, and in general the fix for the original bug has worked (thanks again for that!).  Yesterday however, a query/queries seem to have managed to cause a panic in the newly introduced code (stacktrace truncated for brevity)

`2020-12-30T16:22:03.550Z        error   net/http/server.go:3093 http: panic serving 10.11.25.8:43426: runtime error: slice bounds out of range [70:69]
goroutine 5287869 [running]:
net/http.(*conn).serve.func1(0xc0410b50e0)
        net/http/server.go:1801 +0x147
panic(0xc383e0, 0xc03b45cda0)
        runtime/panic.go:975 +0x47a
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/graphite.getRegexpStringForQuery(0xc03ebc22d0, 0x45, 0xc01cd9002e, 0x11364c0, 0xc03ebc2269)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/graphite/metrics_api.go:403 +0x1010
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/graphite.getRegexpForQuery(0xc03ebc22d0, 0x45, 0x2e, 0x0, 0x0, 0x0)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/graphite/metrics_api.go:341 +0x125
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/graphite.metricsFind(0x0, 0x176b474cff8, 0x0, 0x0, 0xc03ebc2230, 0x4f, 0x12e, 0x5feca949, 0x6fc23ac00, 0xc85c7f, ...)
     ......
net/http.serverHandler.ServeHTTP(0xc01043e000, 0xd5a480, 0xc045cda380, 0xc0001f6e00)`

My first question is - do you have any suggestion for how I could track down what the query was that broke it, using logging/metrics within Victoria Metrics itself? If that isn't possible, we can investigate additional logging in the application a layer above that sends the requests.

In terms of the after-effect of the panic regards to service availability - the main process didn't die, but nearly 10 minutes later the application did seem to read a large 10GB chunk of data from disk. Likewise, the reported amount of memory used by the process doubled from ~7GB to 14GB.  Query durations were increased for around a 10-15minute period of time, presumably due to the increased I/O.  My second question is, does it make sense that these after-effects are directly related to the original panic exceptions?  I was a bit confused by the time delay between the two events

Many thanks in advance for any help,

Cheers

**Comment by marcuzh** [comment]
Hi @valyala

Having looked a bit harder - we think there were two (unrelated) things going on.

Regards to the panic - a malformed query made it through from grafana. e.g. **foo.{bar,** which is missing the closing curly brace.  I don't believe the panic does anything nefarious, so I could understand they may not be a cause for defensive programming - validating the query first.

We think the big spike in memory and disk I/O was just an unfortunately timed large query, attempting to read hundreds of thousands of series.

Cheers

**Comment by valyala** [comment]
@marcuzh , could you upgrade to the latest available version of VictoriaMetrics - `v1.57.1` at the moment - and verify whether the malformed query leads to a panic there? There were a few changes since then in the logic that handles Graphite queries, so it is likely the panic could be fixed there. See https://github.com/VictoriaMetrics/VictoriaMetrics/releases .

If the panic still persists, I'd recommend opening a new issue.
